compiler/p/codegen/OMRTreeEvaluator.cpp

```diff
@@ -3134,6 +3134,125 @@ static inline void loadArrayCmpSources(TR::Node *node, TR::InstOpCode::Mnemonic
+
+   TR::Register *vec0Reg = cg->allocateRegister(TR_VRF);
+   TR::Register *vec1Reg = cg->allocateRegister(TR_VRF);
+   TR::Register *condReg = cg->allocateRegister(TR_CCR);
+
+   TR::LabelSymbol *startLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol *loopStartLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol *residueStartLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol *endLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol *resultLabel = generateLabelSymbol(cg);
+
+   generateLabelInstruction(cg, TR::InstOpCode::label, node, startLabel);
+   startLabel->setStartInternalControlFlow();
+
+   generateTrg1ImmInstruction(cg, TR::InstOpCode::li, node, indexReg, 0);
+   generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::cmpi4, node, condReg, tempReg, 16);
+
+   // We don't need length anymore as we can calculate the appropriate index by using indexReg and the remainder
+   generateTrg1Src1Imm2Instruction(cg, TR::InstOpCode::rlwinm, node, returnReg, tempReg, 0, 0xF);
+   generateConditionalBranchInstruction(cg, TR::InstOpCode::blt, node, residueStartLabel, condReg);
+
+   generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::srawi, node, tempReg, tempReg, 4);
+   generateSrc1Instruction(cg, TR::InstOpCode::mtctr, node, tempReg);
+
+   generateLabelInstruction(cg, TR::InstOpCode::label, node, loopStartLabel);
+
+   // main-loop
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::lxvb16x, node, vec0Reg, indexReg, src1AddrReg);
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::lxvb16x, node, vec1Reg, indexReg, src2AddrReg);
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::vcmpneb_r, node, vec0Reg, vec0Reg, vec1Reg);
+
+   generateConditionalBranchInstruction(cg, TR::InstOpCode::bne, node, resultLabel, condReg);
+   generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::addi, node, indexReg, indexReg, 16);
+
+   generateConditionalBranchInstruction(cg, TR::InstOpCode::bdnz, node, loopStartLabel, condReg);
+   // main-loop end
+
+   // residue start
+   generateLabelInstruction(cg, TR::InstOpCode::label, node, residueStartLabel);
+
+   generateShiftLeftImmediateLong(cg, node, temp2Reg, returnReg, 56);
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::add, node, tempReg, src1AddrReg, indexReg);
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::lxvll, node, vec0Reg, tempReg, temp2Reg);
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::add, node, tempReg, src2AddrReg, indexReg);
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::lxvll, node, vec1Reg, tempReg, temp2Reg);
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::vcmpneb, node, vec0Reg, vec0Reg, vec1Reg);
+   // residue end
+
+   // result
+   generateLabelInstruction(cg, TR::InstOpCode::label, node, resultLabel);
+
+   generateTrg1Src1Instruction(cg, TR::InstOpCode::vclzlsbb, node, tempReg, vec0Reg);
+
+   if (!node->isArrayCmpLen())
+      {
+      generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::addi, node, returnReg, returnReg, -1);
+      }
+
```

Comment:
the logic here is wrong to me:  when the control flow comes from the bne, the remainder should be irrelevant.  returnReg (the remainder) should never appear in the calculation in that case (being masked out or something). as your code is, it still can come into the equation under certain situations.  thus, it appears wrong to me.

Reply:
I agree that in case of bne, the remainder is irrelevant. And this is by the fact that in case of any mismatch the `matched-byte-count` will always be less that the remainder. And the remainder is masked off by the later -
```
generateTrg1Src2Instruction(cg, TR::InstOpCode::cmp4, node, condReg, tempReg, returnReg);
generateTrg1Src3Instruction(cg, TR::InstOpCode::isellt, node, returnReg, tempReg, returnReg, condReg);
```
as the `matched-byte-count`(tempReg) will be less than the `remainder`(returnReg).
For unmatched case, the remainder will only be used when `matched-byte-count + 1 == remainder` which in case should be fine as `matched-byte-count` and `remainder - 1` will have same value.

Reply:
>  in case of any mismatch the matched-byte-count will always be less that the remainder

how does this make any sense?  the matched-byte-count and the remainder are totally independent two quantities in this case.  how did you conclude one is definitely less than the other.

Reply:
You are correct, I made the mistake because in my previous implementation the remainder was actually the remaining length. Thank you.

Reply:
you are making changes more complicated than necessary.  compare match-byte-count to 16 is working for both cases.

plus, you can do returnReg-1 proactively too, and only select returnReg on condition of previous 16-comparison being equal.